langchain/tools/structured.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import annotations
+
+import logging
+from abc import abstractmethod
+from inspect import Parameter, signature
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    Dict,
+    Generic,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from pydantic import BaseModel, Extra, Field, create_model, validate_arguments
+from pydantic.generics import GenericModel
+
+from langchain.callbacks import get_callback_manager
+from langchain.callbacks.base import BaseCallbackManager
+
+logger = logging.getLogger(__name__)
+
+OUTPUT_T = TypeVar("OUTPUT_T")
+
+
+class BaseStructuredTool(
+    GenericModel,
+    Generic[OUTPUT_T],
+    BaseModel,
+):
+    """Parent class for all structured tools."""
+
+    name: str
+    description: str
+    return_direct: bool = False
+    verbose: bool = False
+    callback_manager: BaseCallbackManager = Field(default_factory=get_callback_manager)
+    args_schema: Type[BaseModel]  # :meta private:
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+        arbitrary_types_allowed = True
+
+    @property
+    def args(self) -> Dict:
+        """Return the JSON schema for the tool's args."""
+        schema = self.args_schema.schema()
+        result = {"properties": schema["properties"]}
```

Comment:
In this PR, since we make the schema mandatory, I actually would prefer to remove this property entirely. If the function they use includes a nested pydantic schema, then the schema will include `#ref`'s , and I'd rather not worry about that right now.